gpii/node_modules/gpii-oauth2/gpii-oauth2-utilities/src/OAuth2Utilities.js

```diff
@@ -93,3 +59,35 @@ gpii.oauth2.mapPromiseToResponse = function (promise, response) {
     });
 };
 
+/**
+ * Returns the current time in a human readable string that also naturally sort in chronological order.
+ * See http://www.ecma-international.org/ecma-262/5.1/#sec-15.9.5.43
+ * @return {String} The current time in ISO string format.
+ */
+gpii.oauth2.getCurrentTimestamp = function () {
+    return new Date().toISOString();
+};
+
+/**
+ * Calculate the timestamp of currentTime + expiresIn.
+ * @param expiresIn {Number} The number of seconds that the expiration will occur.
+ * @return {Number} The number of seconds that the expiration will occur.
+ */
+gpii.oauth2.getTimestampExpires = function (expiresIn) {
+    return new Date(new Date().getTime() + expiresIn * 1000).toISOString();
+};
+
+/**
+ * Compare the current time with the expiresIn time and return the number of seconds that the expiration will occur.
+ * @param timestampExpires {String} A string in the format returned by Date().toISOString()
+ * @return {Number} The number of seconds that the expiration will occur. Return 0 if the given timestampExpires < the current timestamp.
+ */
+gpii.oauth2.getExpiresIn = function (timestampExpires) {
+    if (!timestampExpires) {
+        return undefined;
+    }
+
+    var currentTimeInMsec = new Date().getTime();
+    var expiresTimeInMsec = new Date(timestampExpires).getTime();
+    return expiresTimeInMsec > currentTimeInMsec ? Math.round((expiresTimeInMsec - currentTimeInMsec) / 1000) : 0;
+};
```

Comment:
I'm wondering a little about the design of the functions here that process dates and testability. We don't have a test for `gpii.oauth2.getTimestampExpires` and we do have a test for `gpii.oauth2.getExpiresIn` here:

https://github.com/cindyli/universal/blob/GPII-2435/gpii/node_modules/gpii-oauth2/gpii-oauth2-utilities/test/js/OAuth2UtilitiesTests.js#L131-L154

There's a low probability that the last test linked above would fail as both it and the function under test create dates. It would fail in the case that the test date creation and the creation of the date in the function are on either side of a second change boundary. This will be most likely perhaps on a loaded CI server.

Do we have other cases in the codebase where we have date processing? If so what approach have we taken there?

In other systems that I have worked on, I have seen an approach used where all functions that process dates are provided dates as function arguments by the code using the function and never make their own. This approach makes testing reliable and predictable.

@amb26 Would love to get your thoughts here

Reply:
Yes, I agree that we need to mock out date provision. It is the proper solution to problems of this kind that me and @stegru have run into from time to time, which we have generally copped out of.